File: visione/services/core/src/main/java/it/cnr/isti/visione/logging/DRESClient.java
package it.cnr.isti.visione.logging;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.net.ConnectException;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.util.List;

import org.openapitools.client.model.*;

import com.google.gson.Gson;

import dev.dres.ApiClient;
import dev.dres.ApiException;
import dev.dres.client.EvaluationClientApi;
import dev.dres.client.LogApi;
import dev.dres.client.SubmissionApi;
import dev.dres.client.UserApi;
import it.cnr.isti.visione.services.Settings;

public class DRESClient {

	private UserApi userApi;
	private EvaluationClientApi runInfoApi;
	private SubmissionApi submissionApi;
	private LogApi logApi;
	private String sessionId;
	private Gson gson = new Gson();
	private File LOGGING_FOLDER_DRES;


	public static void main(String[] args) {
		String videoId = "06809";
		double timestamp = 1753.385;
		long time = (long) (timestamp*1000); //Tools.convertTimeToVBSFormat(timestamp);
		int frameNumber = 35460;
		System.out.println(time);
		//time = "00:29:13:10";

		DRESClient client = new DRESClient();

//		client.dresSubmitResultByFrameNumber(videoId, frameNumber);
		try {
			client.dresSubmitResultByTime(videoId, time, time);
		} catch (ApiException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}

	public DRESClient() {

		LOGGING_FOLDER_DRES = new File(Settings.LOG_FOLDER_DRES);

		ApiClient client = new ApiClient().setBasePath(Settings.SUBMIT_SERVER);

        //initialize user api client
        userApi = new UserApi(client);

        //initialize evaluation run info client
        runInfoApi = new EvaluationClientApi(client);

        //initialize submission api client
        submissionApi = new SubmissionApi(client);

        //initialize logging api client
        logApi = new LogApi(client);

        System.out.println("Trying to log in to '" + Settings.SUBMIT_SERVER + "' with user '" + Settings.SUBMIT_USER + "'");

        //login request
        ApiUser login = null;
        try {
            login = userApi.postApiV2Login(new LoginRequest().username(Settings.SUBMIT_USER).password(Settings.SUBMIT_PWD));
            System.out.println("-->DRES:login successful");
            System.out.println("user: " + login.getUsername());
            System.out.println("role: " + login.getRole().getValue());
            System.out.println("session: " + login.getSessionId());

            //store session token for future requests
            sessionId = login.getSessionId();
            saveSessionInfo(sessionId, login.getUsername(), Settings.MEMBER_ID, login.getRole().getValue());
        } catch (ApiException e) {

            if (e.getCause() instanceof ConnectException) {
                System.err.println("-->DRES: Could not connect to " + Settings.SUBMIT_SERVER + ", exiting");
            } else {
                System.err.println("-->DRES: Error during login request: '" + e.getMessage() + "', exiting");
            }
        } catch (IOException e) {
            System.err.println("Error, unable to write DRES logging info file for sessionId " + sessionId);
			e.printStackTrace();
		}
	}

	public String getSessionId() {
		return sessionId;
	}

	public String getEvaluationId() {
		List<ApiClientEvaluationInfo> currentRuns;
		try {
		currentRuns = runInfoApi.getApiV2ClientEvaluationList(sessionId);
		} catch (Exception e) {
			System.out.println("-->DRES: Error during request: '" + e.getMessage() + "', exiting");
		return e.getMessage();
		}

		System.out.println("Found " + currentRuns.size() + " ongoing evaluation runs");

		for (ApiClientEvaluationInfo run : currentRuns) {
		System.out.println(run.getName() + " (" + run.getId() + "): " + run.getStatus());
		if (run.getTemplateDescription() != null) {
			System.out.println(run.getTemplateDescription());
		}
		System.out.println();
		}

		String evaluationId=null;
		ApiClientEvaluationInfo firstRun= currentRuns.stream().filter(evaluation -> evaluation.getStatus() == ApiEvaluationStatus.ACTIVE).findFirst().orElseGet(null);
		if(firstRun!=null){
			evaluationId=firstRun.getId();
		}
		
		//print evaluation id
		//evaluationId="5ffa5b86-a0d4-47cf-93cb-cb320180cd5e"; 
		System.out.println("-->DRES: Using evaluationId: " + evaluationId);
		return evaluationId;
		}



		public String dresSubmitLSC(String imagefilename) throws ApiException {//used in KIS and AVS Tasks in LSC
			System.out.println("Submission to DRES (SessionId: " + sessionId + ")");
			String evaluationId = getEvaluationId();
			System.out.println("Submitting " + imagefilename );
			SuccessfulSubmissionsStatus submissionResponse = null;
			try {
				submissionResponse = submissionApi.postApiV2SubmitByEvaluationId(evaluationId,
						new ApiClientSubmission().addAnswerSetsItem(
							new ApiClientAnswerSet().addAnswersItem(
								new ApiClientAnswer()
									.mediaItemName(imagefilename)
									.start((long) 0) //start time in milliseconds
									.end((long) 0)
							)
						), sessionId);
			} catch (ApiException e) {
				String message = "";
				ErrorMessages errorMessage = gson.fromJson(e.getResponseBody(), ErrorMessages.class);
				switch (e.getCode()) {
					case 401: {
						message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ". There was an authentication error during the submission. Check the session id.";
						System.err.println(message);
						throw new ApiException(message);
					}
					case 404: {
						message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ". There is currently no active task which would accept submissions.";
						System.err.println(message);
						break;
					}
					case 412: {
						message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ". The submission was rejected by the server";
						System.err.println(message);
						break;
					}
					default: {
						message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ".Something unexpected went wrong during the submission";
						System.err.println(message);                }
				}
				return message;
			}
	
			if (submissionResponse  != null && submissionResponse.getStatus()) {
				System.out.println("-->DRES: The submission was successfully sent to the server.");
			}
			return submissionResponse.getDescription();
		}

	public String dresSubmitResultByTime(String video, long startTime, long endTime) throws ApiException {//used in KIS and AVS Tasks
        System.out.println("Submission to DRES (SessionId: " + sessionId + ")");
		String evaluationId = getEvaluationId();
		System.out.println("Submitting " + video + " @ start: " + startTime + " - end:"+endTime);
		SuccessfulSubmissionsStatus submissionResponse = null;
        try {
        	submissionResponse = submissionApi.postApiV2SubmitByEvaluationId(evaluationId,
                    new ApiClientSubmission().addAnswerSetsItem(
                        new ApiClientAnswerSet().addAnswersItem(
                            new ApiClientAnswer()
                                .mediaItemName(video)
                                .start(startTime) //start time in milliseconds
								.end(endTime)
                        )
                    ), sessionId);
        } catch (ApiException e) {
        	String message = "";
            ErrorMessages errorMessage = gson.fromJson(e.getResponseBody(), ErrorMessages.class);
            switch (e.getCode()) {
                case 401: {
                	message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ". There was an authentication error during the submission. Check the session id.";
                    System.err.println(message);
                    throw new ApiException(message);
                }
                case 404: {
                    message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ". There is currently no active task which would accept submissions.";
                    System.err.println(message);
                    break;
                }
                case 412: {
                	message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ". The submission was rejected by the server";
                    System.err.println(message);
                    break;
                }
                default: {
                	message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ".Something unexpected went wrong during the submission";
                    System.err.println(message);                }
            }
            return message;
        }

        if (submissionResponse  != null && submissionResponse.getStatus()) {
            System.out.println("-->DRES: The submission was successfully sent to the server.");
        }
        return submissionResponse.getDescription();
	}

	public String dresSubmitTextAnswer(String userAnswer) throws ApiException { //used in Question Aswering Tasks
        System.out.println("Submission to DRES (SessionId: " + sessionId + ")");
		String evaluationId = getEvaluationId();
        System.out.println("Submitting userAnswer");
		SuccessfulSubmissionsStatus submissionResponse = null;
        try {
        	submissionResponse = submissionApi.postApiV2SubmitByEvaluationId(evaluationId,
                    new ApiClientSubmission().addAnswerSetsItem(
                        new ApiClientAnswerSet().addAnswersItem(
                            new ApiClientAnswer()
                                .text(userAnswer)
                        )
                    ), sessionId);
        } catch (ApiException e) {
        	String message = "";
            ErrorMessages errorMessage = gson.fromJson(e.getResponseBody(), ErrorMessages.class);
            switch (e.getCode()) {
                case 401: {
                	message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ". There was an authentication error during the submission. Check the session id.";
                    System.err.println(message);
                    throw new ApiException(message);
                }
                case 404: {
                    message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ". There is currently no active task which would accept submissions.";
                    System.err.println(message);
                    break;
                }
                case 412: {
                	message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ".";
                    System.err.println(message);
                    break;
                }
                default: {
                	message = "-->DRES: Error " + e.getCode() + " " +  e.getMessage() + ","  + errorMessage.description + ".Something unexpected went wrong during the submission";
                    System.err.println(message);                }
            }
            return message;
        }

        if (submissionResponse  != null && submissionResponse.getStatus()) {
            System.out.println("-->DRES: The submission was successfully sent to the server.");
        }
        return submissionResponse.getDescription();
	}

	// public void dresSubmitQuery(QueryEventLog eventLog) throws KeyManagementException, NoSuchAlgorithmException, NumberFormatException {
	// 	DresQueryLogging queryLogging = new DresQueryLogging(eventLog);
    //     new Thread(queryLogging).start();
	// }

	public void dresSubmitLog(QueryResultLog resultLog) throws KeyManagementException, NoSuchAlgorithmException, NumberFormatException {
		DresResultsLogging resultLogging = new DresResultsLogging(resultLog, getSessionId(), getEvaluationId());
        new Thread(resultLogging).start();
	}

	public SuccessStatus dresLogout() throws ApiException {
		 SuccessStatus logout = null;

		try {
		logout = userApi.getApiV2Logout(sessionId);
		} catch (ApiException e) {
			System.err.println("-->DRES: Error during request: '" + e.getMessage() + "'");
		}

		if (logout != null && logout.getStatus()) {
			System.out.println("-->DRES: Successfully logged out");
		}

        return logout;
	}



	public synchronized void saveSessionInfo(String sessionId, String username, String memberID, String role) throws IOException {
		if (!LOGGING_FOLDER_DRES.exists())
			LOGGING_FOLDER_DRES.mkdir();
		String text = "{\"session_id\": \"" + sessionId + "\", \"username\": \"" + username + "\", \"memberID\": \"" + memberID  + "\" , \"role\": \"" + role + "\" }";
		long time = System.currentTimeMillis();
		try (BufferedWriter writer = new BufferedWriter(new FileWriter(new File(LOGGING_FOLDER_DRES, "session_" + time + ".json")))) {
			writer.write(text.toString());
		}

	}

	// private class DresQueryLogging implements Runnable { //FIXME

	// 	QueryEventLog eventLog;

	// 	public DresQueryLogging(QueryEventLog eventLog) {
	// 		this.eventLog = eventLog;
	// 	}

	// 	public String submitQuery(QueryEventLog eventLog) throws KeyManagementException, NoSuchAlgorithmException, NumberFormatException {
	//         SuccessStatus res = null;
	// 		try {
	//            res = logApi.postApiV2LogQuery(sessionId, eventLog);
	//         } catch (ApiException e) {
	//         	String message = "Error during request: '" + e.getMessage() + "'";
	//             System.err.println(message);
	//             return message;
	//         }
	//         return res.getDescription();
	// 	}

	//     @Override
	//     public void run() {
	//     	try {
	// 			String res = submitQuery(eventLog);
	// 	        System.out.println(res);
	// 		} catch (KeyManagementException | NumberFormatException | NoSuchAlgorithmException e) {
	// 			// TODO Auto-generated catch block
	// 			e.printStackTrace();
	// 		}
	//     }
	// }


	private class DresResultsLogging implements Runnable {

		QueryResultLog resultsLog;
		String sessionId;
		String evaluationId;

		public DresResultsLogging(QueryResultLog resultsLog, String sessionId,String evaluationId) {
			this.resultsLog = resultsLog;
			this.sessionId = sessionId;
			this.evaluationId = evaluationId;
		}

		public String submitResults() throws KeyManagementException, NoSuchAlgorithmException, NumberFormatException {
	        SuccessStatus res = null;
			try {
	           //res = logApi.postApiV2LogResult(sessionId, resultsLog);
			   	res= logApi.postApiV2LogResultByEvaluationId(
                        evaluationId,
                        sessionId,
						resultsLog
                );
				//System.out.println("--->LogResults sent to DRES");
	        } catch (ApiException e) {
	        	String message = "-->DRES: Error during request: '" + e.getMessage() + "'";
	            System.err.println(message);
	            return message;
	        }
	        return res.getDescription();
		}

	    @Override
	    public void run() {
	    	try {
				String res = submitResults();
		        System.out.println("-->DRES: "+res);
			} catch (KeyManagementException | NumberFormatException | NoSuchAlgorithmException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
	    }
	}

}

File: visioneUI/src/services/dresClient.ts
import {
  ApiEvaluationStatus,
  Configuration,
  EvaluationClientApi,
  LogApi,
  type QueryResultLog,
  type SuccessStatus,
  SubmissionApi,
  type SuccessfulSubmissionsStatus,
  type ApiClientEvaluationInfo,
  type ApiUser,
  UserApi,
  type LoginRequest,
  ResponseError
} from '$lib/dres-client';

export type DresClientOptions = {
  basePath: string;
  username: string;
  password: string;
  memberId?: string;
  persistSessionInfo?: boolean;
};

export type DresSessionInfo = {
  sessionId: string;
  username: string;
  memberId?: string;
  role?: string;
  createdAt: string;
};

function normalizeBasePath(value: string): string {
  return value.trim().replace(/\/+$/, '');
}

export class DresClientError extends Error {
  statusCode?: number;
  description?: string;

  constructor(message: string, statusCode?: number, description?: string) {
    super(message);
    this.name = 'DresClientError';
    this.statusCode = statusCode;
    this.description = description;
  }
}

export class DresClient {
  private readonly userApi: UserApi;
  private readonly evaluationClientApi: EvaluationClientApi;
  private readonly submissionApi: SubmissionApi;
  private readonly logApi: LogApi;
  private readonly options: DresClientOptions;

  private sessionId: string | null = null;

  constructor(options: DresClientOptions) {
    this.options = {
      ...options,
      basePath: normalizeBasePath(options.basePath)
    };

    const configuration = new Configuration({
      basePath: this.options.basePath
    });

    this.userApi = new UserApi(configuration);
    this.evaluationClientApi = new EvaluationClientApi(configuration);
    this.submissionApi = new SubmissionApi(configuration);
    this.logApi = new LogApi(configuration);
  }

  getSessionId(): string | null {
    return this.sessionId;
  }

  private logDresResponse(operation: string, response: unknown): void {
    console.info(`[DRES] ${operation} response`, response);
  }

  private logDresError(operation: string, error: unknown): void {
    console.error(`[DRES] ${operation} error`, error);
  }

  private normalizeMediaItemName(mediaItemName: string): string {
    return String(mediaItemName).replace(/\.[^./\\]+$/i, '');
  }

  async login(): Promise<ApiUser> {
    const payload: LoginRequest = {
      username: this.options.username,
      password: this.options.password
    };

    try {
      const user = await this.userApi.postApiV2Login({ loginRequest: payload });
      if (!user?.sessionId) {
        throw new DresClientError('Login succeeded but sessionId is missing in the response.');
      }

      this.sessionId = user.sessionId;
      this.saveSessionInfo({
        sessionId: user.sessionId,
        username: user.username ?? this.options.username,
        memberId: this.options.memberId,
        role: user.role,
        createdAt: new Date().toISOString()
      });

      this.logDresResponse('login', user);

      return user;
    } catch (error) {
      this.logDresError('login', error);
      throw await this.mapApiError(error, 'Error during DRES login');
    }
  }

  async getEvaluationId(): Promise<string> {
    const session = this.requireSession();

    let runs: ApiClientEvaluationInfo[];
    try {
      runs = await this.evaluationClientApi.getApiV2ClientEvaluationList({ session });
      this.logDresResponse('getEvaluationList', runs);
    } catch (error) {
      this.logDresError('getEvaluationList', error);
      throw await this.mapApiError(error, 'Error while fetching evaluations');
    }

    const active = runs.find((run) => run.status === ApiEvaluationStatus.Active);
    if (!active?.id) {
      throw new DresClientError('No ACTIVE evaluation is available for the current session.', 404);
    }

    return active.id;
  }

  async listEvaluations(): Promise<ApiClientEvaluationInfo[]> {
    const session = this.requireSession();

    try {
      const response = await this.evaluationClientApi.getApiV2ClientEvaluationList({ session });
      this.logDresResponse('listEvaluations', response);
      return response;
    } catch (error) {
      this.logDresError('listEvaluations', error);
      throw await this.mapApiError(error, 'Error while fetching evaluations');
    }
  }

  async submitResultByTime(
    videoId: string,
    startMs: number,
    endMs: number = startMs,
    evaluationId: string
  ): Promise<SuccessfulSubmissionsStatus> {
    const session = this.requireSession();
    const normalizedMediaItemName = this.normalizeMediaItemName(videoId);
    const apiClientSubmission = {
      answerSets: [
        {
          answers: [
            {
              mediaItemName: normalizedMediaItemName,
              start: Math.floor(startMs),
              end: Math.floor(endMs)
            }
          ]
        }
      ]
    };

    console.info('[DRES] submitResultByTime payload', {
      evaluationId,
      session,
      apiClientSubmission
    });

    try {
      const response = await this.submissionApi.postApiV2SubmitByEvaluationId({
        evaluationId,
        session,
        apiClientSubmission
      });
      this.logDresResponse('submitResultByTime', response);
      return response;
    } catch (error) {
      this.logDresError('submitResultByTime', error);
      throw await this.mapSubmissionError(error);
    }
  }

  async submitResultByImgId(
    imageFilename: string,
    evaluationId: string
  ): Promise<SuccessfulSubmissionsStatus> {
    const session = this.requireSession();
    const normalizedMediaItemName = this.normalizeMediaItemName(imageFilename);
    const apiClientSubmission = {
      answerSets: [
        {
          answers: [
            {
              mediaItemName: normalizedMediaItemName
            }
          ]
        }
      ]
    };

    console.info('[DRES] submitResultByImgId payload', {
      evaluationId,
      session,
      apiClientSubmission
    });

    try {
      const response = await this.submissionApi.postApiV2SubmitByEvaluationId({
        evaluationId,
        session,
        apiClientSubmission
      });
      this.logDresResponse('submitResultByImgId', response);
      return response;
    } catch (error) {
      this.logDresError('submitResultByImgId', error);
      throw await this.mapSubmissionError(error);
    }
  }

  async submitTextAnswer(text: string, evaluationId: string): Promise<SuccessfulSubmissionsStatus> {
    const session = this.requireSession();
    const apiClientSubmission = {
      answerSets: [
        {
          answers: [
            {
              text
            }
          ]
        }
      ]
    };

    console.info('[DRES] submitTextAnswer payload', {
      evaluationId,
      session,
      apiClientSubmission
    });

    try {
      const response = await this.submissionApi.postApiV2SubmitByEvaluationId({
        evaluationId,
        session,
        apiClientSubmission
      });
      this.logDresResponse('submitTextAnswer', response);
      return response;
    } catch (error) {
      this.logDresError('submitTextAnswer', error);
      throw await this.mapSubmissionError(error);
    }
  }

  async submitResultLog(resultsLog: QueryResultLog, evaluationId: string): Promise<SuccessStatus> {
    const session = this.requireSession();

    try {
      const response = await this.logApi.postApiV2LogResultByEvaluationId({
        evaluationId,
        session,
        queryResultLog: resultsLog
      });
      this.logDresResponse('submitResultLog', response);
      return response;
    } catch (error) {
      this.logDresError('submitResultLog', error);
      throw await this.mapApiError(error, 'Error while sending result log');
    }
  }

  async logout(): Promise<SuccessStatus> {
    const session = this.requireSession();

    try {
      const response = await this.userApi.getApiV2Logout({ session });
      this.sessionId = null;
      this.logDresResponse('logout', response);
      return response;
    } catch (error) {
      this.logDresError('logout', error);
      throw await this.mapApiError(error, 'Error during DRES logout');
    }
  }

  private requireSession(): string {
    if (!this.sessionId) {
      throw new DresClientError('DRES session is missing. Run login() first.', 401);
    }
    return this.sessionId;
  }

  private saveSessionInfo(info: DresSessionInfo): void {
    if (!this.options.persistSessionInfo) return;
    if (typeof window === 'undefined') return;

    const key = `dres:session:${Date.now()}`;
    localStorage.setItem(key, JSON.stringify(info));
  }

  private async mapSubmissionError(error: unknown): Promise<DresClientError> {
    if (error instanceof ResponseError) {
      const statusCode = error.response.status;
      const description = await this.extractErrorDescription(error.response);

      if (statusCode === 401) {
        return new DresClientError(
          `DRES 401: authentication failed during submission. ${description ?? ''}`.trim(),
          statusCode,
          description
        );
      }

      if (statusCode === 404) {
        return new DresClientError(
          `DRES 404: no active task accepts submissions. ${description ?? ''}`.trim(),
          statusCode,
          description
        );
      }

      if (statusCode === 412) {
        return new DresClientError(
          `DRES 412: submission rejected by server. ${description ?? ''}`.trim(),
          statusCode,
          description
        );
      }

      return new DresClientError(
        `DRES ${statusCode}: error during submission. ${description ?? ''}`.trim(),
        statusCode,
        description
      );
    }

    if (error instanceof Error) {
      return new DresClientError(error.message);
    }

    return new DresClientError('Unknown error during submission');
  }

  private async mapApiError(error: unknown, contextMessage: string): Promise<DresClientError> {
    if (error instanceof ResponseError) {
      const statusCode = error.response.status;
      const description = await this.extractErrorDescription(error.response);
      return new DresClientError(
        `${contextMessage} (HTTP ${statusCode})${description ? `: ${description}` : ''}`,
        statusCode,
        description
      );
    }

    if (error instanceof Error) {
      return new DresClientError(`${contextMessage}: ${error.message}`);
    }

    return new DresClientError(contextMessage);
  }

  private async extractErrorDescription(response: Response): Promise<string | undefined> {
    try {
      const body = await response.clone().json();
      if (typeof body?.description === 'string') return body.description;
      if (typeof body?.message === 'string') return body.message;
      return undefined;
    } catch {
      try {
        const text = await response.clone().text();
        return text || undefined;
      } catch {
        return undefined;
      }
    }
  }
}


export function createDresClientFromSettings(settings: {
  dresSubmitServer?: string;
  dresUsername?: string;
  dresPassword?: string;
  dresMemberId?: string;
  dresEnabled?: boolean;
}): DresClient {
  const basePath = normalizeBasePath(settings?.dresSubmitServer ?? '');
  const username = settings?.dresUsername?.trim();
  const password = settings?.dresPassword ?? '';
  const memberId = settings?.dresMemberId?.trim();

  if (!settings?.dresEnabled) {
    throw new DresClientError('DRES is disabled in settings.');
  }

  if (!basePath || !username || !password) {
    throw new DresClientError(
      'Incomplete DRES config. Set submit server, username, and password in settings.'
    );
  }

  return new DresClient({
    basePath,
    username,
    password,
    memberId,
    persistSessionInfo: true
  });
}
